core/src/test/scala/unit/kafka/coordinator/group/GroupCoordinatorConcurrencyTest.scala

```diff
@@ -106,12 +107,29 @@ class GroupCoordinatorConcurrencyTest extends AbstractCoordinatorConcurrencyTest
     }
   }
 
-  def createGroupMembers(groupPrefix: String): Set[GroupMember] = {
-    (0 until nGroups).flatMap { i =>
-      new Group(s"$groupPrefix$i", nMembersPerGroup, groupCoordinator, replicaManager).members
-    }.toSet
+  /**
+   * make CompleteTxnOperation and CommitTxnOffsetsOperation complete atomically since they don't typically overlap.
+   * Otherwise CompleteTxnOperation may see a pending offsetAndMetadata without an appendedBatchOffset.
+   */
+  private val txnLock = new ReentrantLock
+  private val allGroupMembers = mutable.ArrayBuffer[GroupMember]()
+
+  def groupMembers(groupId: String, nMembers: Int, groupCoordinator: GroupCoordinator): Seq[GroupMember] = {
+    val groupPartitionId = groupCoordinator.partitionFor(groupId)
+    groupCoordinator.groupManager.addPartitionOwnership(groupPartitionId)
+    val members = (0 until nMembers).map(i => new GroupMember(groupId = groupId,
+      groupPartitionId = groupPartitionId,
+      leader = i == 0,
+      // same producerId to tests more on transactional conflicts.
+      producerId = 1000,
+      txnLock = txnLock))
+    allGroupMembers ++= members
```

Comment:
Since createGroupMembers() is called in multiple tests, it seems we will be accumulating allGroupMembers across tests. That seems unexpected?

Reply:
Junit, by default, creates a new instance for each test case so ```allGroupMembers``` is always new one for each test case.